src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -393,44 +377,191 @@ private static SyntaxList<TypeParameterConstraintClauseSyntax> GetConstraintClau
             }
         }
 
-        private static bool HaveSameConstraints(TypeParameterConstraintClause clause1, TypeParameterConstraintClause clause2)
+        /// <summary>
+        /// Note, only nullability aspects are merged if possible, other mismatches are treated as failures.
+        /// </summary>
+        private ImmutableArray<TypeParameterConstraintClause> MergeConstraintsForPartialDeclarations(ImmutableArray<TypeParameterConstraintClause> constraintClauses, DiagnosticBag diagnostics)
         {
```

Comment:
>{ [](start = 8, length = 1)

Consider moving the assertion that `TypeParameters.Length == constraintClauses.Length` down into this method, where it comes into play.